### ISSUE TITLE
Compile errors when using Xcode 14.3 and C++20

### ISSUE THREAD

**Issue by jwwalker** [OGRECave/ogre-next#370]
#### System Information
- Ogre Version: master branch [c1c24caaae]
- Operating System / Platform: macOS 13.3.1
- RenderSystem: Metal
- GPU: Apple M1

#### Detailed description
If I try to compile the following code

```c++
#include <OgreGpuProgramParams.h>

int main(int argc, const char * argv[])
{
	return 0;
}
```

using Xcode 14.3 set to use c++20, I get compile errors (listed below).  If I use Xcode 14.3 with c++17, or Xcode 14.2 with c++20, there is no error. Possibly this could be an Xcode bug rather than an Ogre bug.

#### Build error log

```
In file included from /Users/jwwalker/TestOgreInclude/TestOgreInclude/main.cpp:8:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:31:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgrePrerequisites.h:32:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/string:551:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/string_view:223:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/functional:515:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__functional/boyer_moore_searcher.h:26:
/Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/vector:540:52: error: arithmetic on a pointer to an incomplete type 'Ogre::GpuProgramParameters_AutoConstantEntry'
        {return static_cast<size_type>(__end_cap() - this->__begin_);}
                                       ~~~~~~~~~~~ ^
/Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/vector:760:56: note: in instantiation of member function 'std::vector<Ogre::GpuProgramParameters_AutoConstantEntry>::capacity' requested here
      __annotate_contiguous_container(data(), data() + capacity(),
                                                       ^
/Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/vector:431:7: note: in instantiation of member function 'std::vector<Ogre::GpuProgramParameters_AutoConstantEntry>::__annotate_delete' requested here
      __annotate_delete();
      ^
In file included from /Users/jwwalker/TestOgreInclude/TestOgreInclude/main.cpp:8:
/Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:1510:9: note: in instantiation of member function 'std::vector<Ogre::GpuProgramParameters_AutoConstantEntry>::~vector' requested here
        ~GpuProgramParameters() {}
        ^
/Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:848:11: note: forward declaration of 'Ogre::GpuProgramParameters_AutoConstantEntry'
    class GpuProgramParameters_AutoConstantEntry;
          ^
In file included from /Users/jwwalker/TestOgreInclude/TestOgreInclude/main.cpp:8:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:31:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgrePrerequisites.h:32:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/string:551:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/string_view:223:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/functional:515:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__functional/boyer_moore_searcher.h:26:
/Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/vector:760:54: error: arithmetic on a pointer to an incomplete type 'const std::vector<Ogre::GpuProgramParameters_AutoConstantEntry>::value_type' (aka 'const Ogre::GpuProgramParameters_AutoConstantEntry')
      __annotate_contiguous_container(data(), data() + capacity(),
                                              ~~~~~~ ^
/Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/vector:431:7: note: in instantiation of member function 'std::vector<Ogre::GpuProgramParameters_AutoConstantEntry>::__annotate_delete' requested here
      __annotate_delete();
      ^
In file included from /Users/jwwalker/TestOgreInclude/TestOgreInclude/main.cpp:8:
/Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:1510:9: note: in instantiation of member function 'std::vector<Ogre::GpuProgramParameters_AutoConstantEntry>::~vector' requested here
        ~GpuProgramParameters() {}
        ^
/Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:848:11: note: forward declaration of 'Ogre::GpuProgramParameters_AutoConstantEntry'
    class GpuProgramParameters_AutoConstantEntry;
          ^
In file included from /Users/jwwalker/TestOgreInclude/TestOgreInclude/main.cpp:8:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:31:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgrePrerequisites.h:32:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/string:551:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/string_view:223:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/functional:515:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__functional/boyer_moore_searcher.h:26:
/Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/vector:833:64: error: arithmetic on a pointer to an incomplete type 'Ogre::GpuProgramParameters_AutoConstantEntry'
        __alloc_traits::destroy(__alloc(), _VSTD::__to_address(--__soon_to_be_end));
                                                               ^ ~~~~~~~~~~~~~~~~
/Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/vector:827:29: note: in instantiation of member function 'std::vector<Ogre::GpuProgramParameters_AutoConstantEntry>::__base_destruct_at_end' requested here
  void __clear() _NOEXCEPT {__base_destruct_at_end(this->__begin_);}
                            ^
/Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/vector:436:9: note: in instantiation of member function 'std::vector<Ogre::GpuProgramParameters_AutoConstantEntry>::__clear' requested here
        __clear();
        ^
In file included from /Users/jwwalker/TestOgreInclude/TestOgreInclude/main.cpp:8:
/Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:1510:9: note: in instantiation of member function 'std::vector<Ogre::GpuProgramParameters_AutoConstantEntry>::~vector' requested here
        ~GpuProgramParameters() {}
        ^
/Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:848:11: note: forward declaration of 'Ogre::GpuProgramParameters_AutoConstantEntry'
    class GpuProgramParameters_AutoConstantEntry;
          ^
In file included from /Users/jwwalker/TestOgreInclude/TestOgreInclude/main.cpp:8:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:31:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgrePrerequisites.h:32:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/string:549:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/memory:848:
/Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__memory/allocator.h:128:58: error: invalid application of 'sizeof' to an incomplete type 'Ogre::GpuProgramParameters_AutoConstantEntry'
            _VSTD::__libcpp_deallocate((void*)__p, __n * sizeof(_Tp), _LIBCPP_ALIGNOF(_Tp));
                                                         ^~~~~~~~~~~
In file included from /Users/jwwalker/TestOgreInclude/TestOgreInclude/main.cpp:8:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:31:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgrePrerequisites.h:32:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/string:534:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__memory/allocate_at_least.h:13:
/Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__memory/allocator_traits.h:282:13: note: in instantiation of member function 'std::allocator<Ogre::GpuProgramParameters_AutoConstantEntry>::deallocate' requested here
        __a.deallocate(__p, __n);
            ^
In file included from /Users/jwwalker/TestOgreInclude/TestOgreInclude/main.cpp:8:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:31:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgrePrerequisites.h:32:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/string:551:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/string_view:223:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/functional:515:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__functional/boyer_moore_searcher.h:26:
/Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/vector:437:25: note: in instantiation of member function 'std::allocator_traits<std::allocator<Ogre::GpuProgramParameters_AutoConstantEntry>>::deallocate' requested here
        __alloc_traits::deallocate(__alloc(), this->__begin_, capacity());
                        ^
In file included from /Users/jwwalker/TestOgreInclude/TestOgreInclude/main.cpp:8:
/Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:1510:9: note: in instantiation of member function 'std::vector<Ogre::GpuProgramParameters_AutoConstantEntry>::~vector' requested here
        ~GpuProgramParameters() {}
        ^
/Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:848:11: note: forward declaration of 'Ogre::GpuProgramParameters_AutoConstantEntry'
    class GpuProgramParameters_AutoConstantEntry;
          ^
In file included from /Users/jwwalker/TestOgreInclude/TestOgreInclude/main.cpp:8:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:31:
In file included from /Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgrePrerequisites.h:32:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/string:551:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/string_view:223:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/functional:515:
In file included from /Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__functional/boyer_moore_searcher.h:26:
/Applications/Xcode14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/vector:537:53: error: arithmetic on a pointer to an incomplete type 'Ogre::GpuProgramParameters_AutoConstantEntry'
        {return static_cast<size_type>(this->__end_ - this->__begin_);}
                                       ~~~~~~~~~~~~ ^
In file included from /Users/jwwalker/TestOgreInclude/TestOgreInclude/main.cpp:8:
/Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:1986:69: note: in instantiation of member function 'std::vector<Ogre::GpuProgramParameters_AutoConstantEntry>::size' requested here
        size_t getAutoConstantCount() const { return mAutoConstants.size(); }
                                                                    ^
/Users/jwwalker/Ogre/ogre-next/OgreMain/include/OgreGpuProgramParams.h:848:11: note: forward declaration of 'Ogre::GpuProgramParameters_AutoConstantEntry'
    class GpuProgramParameters_AutoConstantEntry;
```
[TestOgreInclude.zip](https://github.com/OGRECave/ogre-next/files/11240271/TestOgreInclude.zip)

**Comment by jwwalker** [comment]
After some further narrowing down, I submitted this as a [bug report to the LLVM project](https://github.com/llvm/llvm-project/issues/62166), since it does not seem to happen in compilers other than Clang.

**Comment by darksylinc** [comment]
Ughhhh. What a nasty error.

Now that you simplified the code; it looks like something the compiler/lib should be able to sort out but not easily. There may be some new C++20 guarantees that will bite everyone in the ***.

Let's see what the LLVM devs say.

**Comment by jwwalker** [comment]
The LLVM devs have already closed my issue as a duplicate of [another](https://github.com/llvm/llvm-project/issues/59292).  I don't think they've decided whether it is a bug.  But a few fairly innocuous changes will make OgreMain compile in Xcode 14.3 as C++20.  I'll submit a PR.

**Comment by darksylinc** [comment]
Should be fixed by #370
